File: packages/oxlint-plugin-react-doctor/src/plugin/rules/r3f/three-require-transparent-for-opacity.ts
import { defineRule } from "../../utils/define-rule.js";
import type { EsTreeNode } from "../../utils/es-tree-node.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";
import { getStaticObjectPropertyValue } from "../../utils/get-static-object-property-value.js";
import { isNodeOfType } from "../../utils/is-node-of-type.js";
import { readStaticBoolean } from "../../utils/read-static-boolean.js";
import type { RuleContext } from "../../utils/rule-context.js";
import { getApiReferenceProvenance } from "./utils/get-api-reference-provenance.js";
import { getStaticNumber } from "./utils/get-static-number.js";
import { isMaterialOpacityIgnored } from "./utils/is-material-opacity-ignored.js";
import { isThreeMaterialTransparentByDefault } from "./utils/is-three-material-transparent-by-default.js";
import { isThreeModuleSource } from "./utils/is-three-module-source.js";

const readOptionalStaticBoolean = (
  expression: EsTreeNode | null | undefined,
): boolean | null | undefined =>
  expression === undefined ? undefined : readStaticBoolean(expression);

const readOptionalStaticNumber = (
  expression: EsTreeNode | null | undefined,
  context: RuleContext,
): number | null | undefined =>
  expression === undefined
    ? undefined
    : expression === null
      ? null
      : getStaticNumber(expression, context.scopes);

export const threeRequireTransparentForOpacity = defineRule({
  id: "three-require-transparent-for-opacity",
  title: "Three.js material opacity is ignored",
  category: "Correctness",
  severity: "warn",
  recommendation:
    "Enable transparent, alphaHash, or alphaTest when a material opacity below 1 should affect rendering",
  create: (context: RuleContext) => ({
    NewExpression(node: EsTreeNodeOfType<"NewExpression">) {
      const provenance = getApiReferenceProvenance(node.callee, context.scopes);
      const parameters = node.arguments[0];
      if (
        !provenance?.apiName.endsWith("Material") ||
        !isThreeModuleSource(provenance.moduleSource) ||
        isThreeMaterialTransparentByDefault(provenance.apiName) ||
        !parameters ||
        !isNodeOfType(parameters, "ObjectExpression") ||
        parameters.properties.some((property) => isNodeOfType(property, "SpreadElement"))
      ) {
        return;
      }
      const opacityExpression = getStaticObjectPropertyValue(parameters, "opacity");
      if (!opacityExpression) return;
      if (
        !isMaterialOpacityIgnored({
          alphaHash: readOptionalStaticBoolean(
            getStaticObjectPropertyValue(parameters, "alphaHash"),
          ),
          alphaTest: readOptionalStaticNumber(
            getStaticObjectPropertyValue(parameters, "alphaTest"),
            context,
          ),
          opacity: getStaticNumber(opacityExpression, context.scopes),
          transparent: readOptionalStaticBoolean(
            getStaticObjectPropertyValue(parameters, "transparent"),
          ),
        })
      ) {
        return;
      }
      context.report({
        node: opacityExpression,
        message:
          "This material sets opacity below 1 without transparent, alphaHash, or alphaTest, so the opacity does not make the surface translucent",
      });
    },
  }),
});

File: native/oxlint/rules/three-require-transparent-for-opacity.rs
use oxc_ast::AstKind;
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;
use oxc_span::GetSpan;

use crate::{context::LintContext, rule::Rule, AstNode};

const DEFAULT_TRANSPARENT_MATERIAL_CONSTRUCTOR_NAMES: [&str; 5] = [
    "ShadowMaterial",
    "ShadowNodeMaterial",
    "SpriteMaterial",
    "SpriteNodeMaterial",
    "VolumeNodeMaterial",
];

#[derive(Debug, Default, Clone)]
pub struct ThreeRequireTransparentForOpacity;

declare_oxc_lint!(
    /// Require a transparency mode for translucent Three.js material opacity.
    ThreeRequireTransparentForOpacity,
    react_doctor_native,
    correctness,
    version = "0.1.0",
    short_description = "Require a Three.js material transparency mode.",
);

impl Rule for ThreeRequireTransparentForOpacity {
    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
        let AstKind::NewExpression(new_expression) = node.kind() else {
            return;
        };
        let Some(constructor_name) = three_module_api_name(&new_expression.callee, ctx) else {
            return;
        };
        if !constructor_name.ends_with("Material")
            || DEFAULT_TRANSPARENT_MATERIAL_CONSTRUCTOR_NAMES.contains(&constructor_name.as_str())
        {
            return;
        }
        let Some(parameters) = new_expression
            .arguments
            .first()
            .and_then(oxc_ast::ast::Argument::as_expression)
        else {
            return;
        };
        let oxc_ast::ast::Expression::ObjectExpression(object_expression) =
            parameters.get_inner_expression()
        else {
            return;
        };
        if object_expression.properties.iter().any(|property| {
            !matches!(
                property,
                oxc_ast::ast::ObjectPropertyKind::ObjectProperty(_)
            )
        }) {
            return;
        }
        let Some(opacity_expression) = get_static_object_property_value(parameters, "opacity")
        else {
            return;
        };
        let Some(opacity) = resolve_static_number(opacity_expression, ctx) else {
            return;
        };
        if !(0.0..1.0).contains(&opacity) {
            return;
        }
        if let Some(transparent_expression) =
            get_static_object_property_value(parameters, "transparent")
        {
            let Some(is_transparent) = resolve_static_boolean_literal(transparent_expression)
            else {
                return;
            };
            if is_transparent {
                return;
            }
        }
        if let Some(alpha_hash_expression) =
            get_static_object_property_value(parameters, "alphaHash")
        {
            let Some(has_alpha_hash) = resolve_static_boolean_literal(alpha_hash_expression) else {
                return;
            };
            if has_alpha_hash {
                return;
            }
        }
        if let Some(alpha_test_expression) =
            get_static_object_property_value(parameters, "alphaTest")
        {
            let Some(alpha_test) = resolve_static_number(alpha_test_expression, ctx) else {
                return;
            };
            if alpha_test > 0.0 {
                return;
            }
        }
        ctx.diagnostic(
            OxcDiagnostic::warn(
                "This material sets opacity below 1 without transparent, alphaHash, or alphaTest, so the opacity does not make the surface translucent",
            )
            .with_label(opacity_expression.span()),
        );
    }
}

fn resolve_static_boolean_literal(expression: &oxc_ast::ast::Expression) -> Option<bool> {
    let oxc_ast::ast::Expression::BooleanLiteral(boolean_literal) =
        expression.get_inner_expression()
    else {
        return None;
    };
    Some(boolean_literal.value)
}
